designs/carbon-aware.md

```diff
@@ -0,0 +1,158 @@
+# Carbon Aware Karpenter: Optimizing Kubernetes Cluster Autoscaling for Environmental Sustainability
+*Author: [@JacobValdemar](https://github.com/JacobValdemar)*
+
+## Context & Problem
+
+There is a growing concern about environmental sustainability within the context of Kubernetes cluster autoscaling. In multiple comments on [the proposal for moving Karpenter to CNCF](https://github.com/kubernetes/org/issues/4258), the move is backed because of opportunities within environmental sustainability.
+
+I'm currently working on my master's thesis in Computer Engineering (M.Sc.Eng) at Aarhus University located in Denmark. The objective of the thesis is to enable Karpenter to minimize carbon emissions from Kubernetes clusters that run on cloud infrastructure (focus is AWS).
+
+RFC: https://github.com/aws/karpenter/issues/4630
+
+## Fundamentals of Green Software
+I'll try to keep it simple. The reader should be familiar with the following.
+
+A cluster's emissions is made of two elements: embodied emissions and operational emissions. To get the total emissions, one can add them togeather.
+
+- **Embodied carbon emissions**: Manufacturing emissions (CO₂e) amortized over instance lifetime (usually 4 years) divided by how long we use the instance
+- **Operational carbon emissions**: Carbon emitted by electricity grid to produce electricity for the instance in the region where it is used, multiplied by PUE
+
+There is a lot more to Green Software. If you want to learn more, I recommend you to visit [Green Software Practitioner](https://learn.greensoftware.foundation/) (a Green Software Foundation project - an affiliate of the Linux Foundation).
+
+## Solution
+
+### Feature Gate
```

Comment:
Worth exploring an approach where you can configure instance prices separately from Karpenter. This would allow karpenter to be unaware of "carbon aware mode", and you simply provide it the updated prices. One option for prices would be a configmap.

```
priceOverrides:
  - name: m5.large
     zone: us-west-2a
     capacity-type: spot
     price: .001
  - name: m5.xlarge
     zone: us-west-2a
     capacity-type: spot
     price: .002
```

Reply:
Interesting idea - that would generally give users a lot of flexibility. 

However, if it is an override, it would not be possible to combine carbon emission price with actual price as [suggested by sftim](https://github.com/aws/karpenter/pull/4686#discussion_r1335703216) (although that is not a № 1 priority for the feature). Also, it would make it hard get precise operational carbon emissions as the [carbon intensity](https://learn.greensoftware.foundation/carbon-awareness#carbon-intensity) of the energy grid depends on the geographic region of the data center. I'll add it as an option tomorrow (it's 22.30 PM here)

Reply:
I have added it as option 4